src/ReplicatedStorage/Client/Tool/Gun/init.lua

```diff
@@ -85,19 +86,35 @@ function Gun.DrawShot(startPoint: Vector3, endPoint: Vector3, config: table)
 	beam.Anchored = true
 	beam.Locked = true
 	beam.CanCollide = false
+	beam.CanQuery = false
+	beam.CanTouch = false
 	beam.Size = Vector3.new(thickness, thickness, distance)
-	beam.CFrame = CFrame.new(startPoint, endPoint) * CFrame.new(0, 0, -distance / 2)
 
 	local mesh = Instance.new("BlockMesh")
 	mesh.Parent = beam
-	beam.Parent = bulletContainer
 
 	local travelSpeed = (distance/300) / (config.BulletDecoration.BulletSpeed or 1)
 	Debris:AddItem(beam, travelSpeed)
 
 	task.delay(1/30, function()
 		tweenOnce(mesh, TweenInfo.new(travelSpeed - 1/30, Enum.EasingStyle.Quad, Enum.EasingDirection.In), {Scale = Vector3.new(1, 1, 0), Offset = Vector3.new(0, 0, -distance/2)})
 	end)
+
+	-- BeamUpdater updates the beam's cframe for 4 frames so it doesn't look weird when the character moves a lot
+	local renderId = "BeamUpdater-"..os.clock()
+	local endTime = os.clock() + 4*(1/60)
+	local function update()
+		if not beam.Parent or not startAttachment.Parent or os.clock()>endTime then
+			RunService:UnbindFromRenderStep(renderId)
+			return
+		end
+		startPoint = startAttachment.WorldPosition
+		beam.CFrame = CFrame.new(startPoint, endPoint) * CFrame.new(0, 0, -distance / 2 - 0.1)
```

Comment:
CFrame.new(startPoint, endPoint) is deprecated. Use CFrame.lookAt instead. 
https://developer.roblox.com/en-us/api-reference/datatype/CFrame